File: examples/workflows/05-data-pipeline/main.py
#!/usr/bin/env python3
"""
Example 5: Data Pipeline Workflow - Python

Demonstrates a 5-stage data pipeline: Extract → Clean → Enrich → Aggregate → Report
"""

import asyncio
import os
import sys
import time

from axonflow import AxonFlow


async def main():
    agent_url = os.getenv("AXONFLOW_ENDPOINT", os.getenv("AXONFLOW_AGENT_URL", "http://localhost:8080"))
    client_id = os.getenv("AXONFLOW_CLIENT_ID")
    client_secret = os.getenv("AXONFLOW_CLIENT_SECRET")

    if not client_id or not client_secret:
        print("AXONFLOW_CLIENT_ID and AXONFLOW_CLIENT_SECRET must be set")
        sys.exit(1)

    client = AxonFlow(
        endpoint=agent_url,
        client_id=client_id,
        client_secret=client_secret,
    )

    print("✅ Connected to AxonFlow")
    print("🔄 Starting 5-stage data pipeline for customer analytics...\n")

    start_time = time.time()

    try:
        # Stage 1: Extract
        print("📥 Stage 1/5: Extracting customer transaction data...")
        await client.proxy_llm_call(
            user_token=os.getenv("AXONFLOW_USER_TOKEN", "user-123"),
            query="Extract customer purchase data from the last 30 days. Include customer ID, purchase amount, product categories, and timestamps. Simulate 500 customer transactions.",
            request_type="chat",
            context={"provider": "openai"},
        )
        print("✅ Stage 1 complete: Data extracted\n")

        # Stage 2: Transform (Clean & Normalize)
        print("🧹 Stage 2/5: Cleaning and normalizing data...")
        await client.proxy_llm_call(
            user_token=os.getenv("AXONFLOW_USER_TOKEN", "user-123"),
            query="""From the extracted data above, perform the following transformations:
1. Remove duplicate transactions
2. Standardize date formats to ISO 8601
3. Normalize product category names
4. Validate all amounts are positive numbers
5. Flag any anomalies (unusually high amounts)""",
            request_type="chat",
            context={"provider": "openai"},
        )
        print("✅ Stage 2 complete: Data cleaned and normalized\n")

        # Stage 3: Enrich
        print("💎 Stage 3/5: Enriching with customer segments and lifetime value...")
        await client.proxy_llm_call(
            user_token=os.getenv("AXONFLOW_USER_TOKEN", "user-123"),
            query="""Based on the cleaned transaction data:
1. Calculate customer lifetime value (CLV)
2. Segment customers into: VIP (CLV > $5000), Regular ($1000-$5000), New (< $1000)
3. Identify top-spending product categories per segment
4. Calculate average order value per segment""",
            request_type="chat",
            context={"provider": "openai"},
        )
        print("✅ Stage 3 complete: Data enriched with segments and metrics\n")

        # Stage 4: Aggregate
        print("📊 Stage 4/5: Aggregating insights and trends...")
        await client.proxy_llm_call(
            user_token=os.getenv("AXONFLOW_USER_TOKEN", "user-123"),
            query="""Generate aggregated insights:
1. Total revenue by customer segment
2. Growth trends (week-over-week)
3. Top 5 products by revenue
4. Customer churn risk indicators
5. Recommended actions for each segment""",
            request_type="chat",
            context={"provider": "openai"},
        )
        print("✅ Stage 4 complete: Insights aggregated\n")

        # Stage 5: Report
        print("📈 Stage 5/5: Generating executive summary report...")
        report_resp = await client.proxy_llm_call(
            user_token=os.getenv("AXONFLOW_USER_TOKEN", "user-123"),
            query="""Create an executive summary report with:
1. Key metrics (total revenue, customer count, avg order value)
2. Segment analysis
3. Top actionable recommendations
4. Risk alerts (if any)
Format as a concise business report.""",
            request_type="chat",
            context={"provider": "openai"},
        )

        duration = time.time() - start_time

        print("\n📊 CUSTOMER ANALYTICS REPORT")
        print("=" * 60)
        print(report_resp.data)
        print("=" * 60)
        print()
        print(f"⏱️  Pipeline completed in {duration:.1f} seconds")
        print("✅ All 5 stages executed successfully")
        print("💡 Data pipeline: Extract → Clean → Enrich → Aggregate → Report")
    except Exception as e:
        print(f"❌ Pipeline failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/workflows/05-data-pipeline/Main.java
/*
 * Example 5: Data Pipeline Workflow - Java
 *
 * Demonstrates a 5-stage data pipeline: Extract → Clean → Enrich → Aggregate → Report
 */
package com.getaxonflow.examples;

import com.getaxonflow.sdk.AxonFlow;
import com.getaxonflow.sdk.AxonFlowConfig;
import com.getaxonflow.sdk.types.ExecuteQueryRequest;
import com.getaxonflow.sdk.types.ExecuteResponse;

import java.util.Map;
import java.util.Optional;

public class Main {

    public static void main(String[] args) {
        String agentUrl = Optional.ofNullable(System.getenv("AXONFLOW_AGENT_URL"))
                .orElse("http://localhost:8080");
        String clientId = System.getenv("AXONFLOW_CLIENT_ID");
        String clientSecret = System.getenv("AXONFLOW_CLIENT_SECRET");

        if (clientId == null || clientId.isEmpty() || clientSecret == null || clientSecret.isEmpty()) {
            System.err.println("❌ AXONFLOW_CLIENT_ID and AXONFLOW_CLIENT_SECRET must be set");
            System.exit(1);
        }

        AxonFlowConfig config = AxonFlowConfig.builder()
                .endpoint(agentUrl)
                .clientId(clientId)
                .clientSecret(clientSecret)
                .build();

        AxonFlow client = new AxonFlow(config);

        System.out.println("✅ Connected to AxonFlow");
        System.out.println("🔄 Starting 5-stage data pipeline for customer analytics...");
        System.out.println();

        long startTime = System.currentTimeMillis();

        try {
            // Stage 1: Extract
            System.out.println("📥 Stage 1/5: Extracting customer transaction data...");
            client.proxyLLMCall(
                    ExecuteQueryRequest.builder()
                            .userToken("user-123")
                            .query("Extract customer purchase data from the last 30 days. Include customer ID, purchase amount, product categories, and timestamps. Simulate 500 customer transactions.")
                            .requestType("chat")
                            .context(Map.of("model", "gpt-4"))
                            .build()
            );
            System.out.println("✅ Stage 1 complete: Data extracted");
            System.out.println();

            // Stage 2: Transform (Clean & Normalize)
            System.out.println("🧹 Stage 2/5: Cleaning and normalizing data...");
            client.proxyLLMCall(
                    ExecuteQueryRequest.builder()
                            .userToken("user-123")
                            .query("From the extracted data above, perform the following transformations:\n" +
                                    "1. Remove duplicate transactions\n" +
                                    "2. Standardize date formats to ISO 8601\n" +
                                    "3. Normalize product category names\n" +
                                    "4. Validate all amounts are positive numbers\n" +
                                    "5. Flag any anomalies (unusually high amounts)")
                            .requestType("chat")
                            .context(Map.of("model", "gpt-4"))
                            .build()
            );
            System.out.println("✅ Stage 2 complete: Data cleaned and normalized");
            System.out.println();

            // Stage 3: Enrich
            System.out.println("💎 Stage 3/5: Enriching with customer segments and lifetime value...");
            client.proxyLLMCall(
                    ExecuteQueryRequest.builder()
                            .userToken("user-123")
                            .query("Based on the cleaned transaction data:\n" +
                                    "1. Calculate customer lifetime value (CLV)\n" +
                                    "2. Segment customers into: VIP (CLV > $5000), Regular ($1000-$5000), New (< $1000)\n" +
                                    "3. Identify top-spending product categories per segment\n" +
                                    "4. Calculate average order value per segment")
                            .requestType("chat")
                            .context(Map.of("model", "gpt-4"))
                            .build()
            );
            System.out.println("✅ Stage 3 complete: Data enriched with segments and metrics");
            System.out.println();

            // Stage 4: Aggregate
            System.out.println("📊 Stage 4/5: Aggregating insights and trends...");
            client.proxyLLMCall(
                    ExecuteQueryRequest.builder()
                            .userToken("user-123")
                            .query("Generate aggregated insights:\n" +
                                    "1. Total revenue by customer segment\n" +
                                    "2. Growth trends (week-over-week)\n" +
                                    "3. Top 5 products by revenue\n" +
                                    "4. Customer churn risk indicators\n" +
                                    "5. Recommended actions for each segment")
                            .requestType("chat")
                            .context(Map.of("model", "gpt-4"))
                            .build()
            );
            System.out.println("✅ Stage 4 complete: Insights aggregated");
            System.out.println();

            // Stage 5: Report
            System.out.println("📈 Stage 5/5: Generating executive summary report...");
            ExecuteResponse reportResp = client.proxyLLMCall(
                    ExecuteQueryRequest.builder()
                            .userToken("user-123")
                            .query("Create an executive summary report with:\n" +
                                    "1. Key metrics (total revenue, customer count, avg order value)\n" +
                                    "2. Segment analysis\n" +
                                    "3. Top actionable recommendations\n" +
                                    "4. Risk alerts (if any)\n" +
                                    "Format as a concise business report.")
                            .requestType("chat")
                            .context(Map.of("model", "gpt-4"))
                            .build()
            );

            double duration = (System.currentTimeMillis() - startTime) / 1000.0;

            System.out.println();
            System.out.println("📊 CUSTOMER ANALYTICS REPORT");
            System.out.println("============================================================");
            System.out.println(reportResp.getData());
            System.out.println("============================================================");
            System.out.println();
            System.out.printf("⏱️  Pipeline completed in %.1f seconds%n", duration);
            System.out.println("✅ All 5 stages executed successfully");
            System.out.println("💡 Data pipeline: Extract → Clean → Enrich → Aggregate → Report");
        } catch (Exception e) {
            System.err.println("❌ Pipeline failed: " + e.getMessage());
            System.exit(1);
        }
    }
}
